File: GenerarDatos/generar_datos.py
import csv
import random
import sys
import time

def generar_datos(n, archivo_salida="dataset.csv"):
    nombres_h = ["Juan", "Carlos", "Luis", "Pedro", "Miguel", "Andres", "Diego", "Fernando", "Jorge", "Mateo"]
    nombres_m = ["Maria", "Ana", "Laura", "Carmen", "Sofia", "Isabel", "Camila", "Valentina", "Lucia", "Elena"]
    apellidos = ["Garcia", "Martinez", "Rodriguez", "Lopez", "Perez", "Williams", "Gomez", "Silva", "Ruiz", "Gonzalez", "Hernandez", "Diaz"]

    print(f"Preparando la generación de {n:,} registros...")
    start_time = time.time()

    with open(archivo_salida, mode='w', newline='', encoding='utf-8') as file:
        writer = csv.writer(file)
        # Cabecera
        writer.writerow(["ID", "Nombre_Completo", "Sexo", "Edad", "Peso_kg", "Altura_cm"])

        # Para un rendimiento rápido en Python puro al generar millones de datos
        for i in range(1, n + 1):
            sexo = random.choice(["M", "F"])
            nombre = random.choice(nombres_h) if sexo == "M" else random.choice(nombres_m)
            apellido1 = random.choice(apellidos)
            apellido2 = random.choice(apellidos)
            nombre_completo = f"{nombre} {apellido1} {apellido2}"
            
            edad = random.randint(18, 90)
            peso = round(random.uniform(50.0, 110.0), 1)
            altura = random.randint(150, 200)

            writer.writerow([i, nombre_completo, sexo, edad, peso, altura])
            
            # Progress print
            if i % 1000000 == 0:
                print(f"... {i:,} registros generados.")

    elapsed = round(time.time() - start_time, 2)
    print(f"¡Generación completada en {elapsed} segundos! Datos guardados en '{archivo_salida}'.")

if __name__ == "__main__":
    # Valor por defecto: 10,000 registros
    n_registros = 10000
    if len(sys.argv) > 1:
        try:
            n_registros = int(sys.argv[1])
        except ValueError:
            print("Por favor provee un número entero válido.")
            sys.exit(1)
            
    generar_datos(n_registros, "dataset.csv")

File: Ordenamiento/GenerarDatos/generar_datos.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

const char* nombres_h[] = {"Juan", "Carlos", "Luis", "Pedro", "Miguel", "Andres", "Diego", "Fernando", "Jorge", "Mateo"};
const char* nombres_m[] = {"Maria", "Ana", "Laura", "Carmen", "Sofia", "Isabel", "Camila", "Valentina", "Lucia", "Elena"};
const char* apellidos[] = {"Garcia", "Martinez", "Rodriguez", "Lopez", "Perez", "Williams", "Gomez", "Silva", "Ruiz", "Gonzalez", "Hernandez", "Diaz"};

// Función para obtener un double aleatorio en un rango
double rand_uniform(double min, double max) {
    double scale = rand() / (double) RAND_MAX; 
    return min + scale * ( max - min ); 
}

int main(int argc, char *argv[]) {
    long long n = 10000;
    
    // Si el usuario pasa un argumento, lo tomamos como N
    if (argc > 1) {
        n = atoll(argv[1]);
    }
    
    printf("Preparando la generación de %lld registros usando C...\n", n);
    
    FILE *file = fopen("dataset.csv", "w");
    if (file == NULL) {
        printf("Error fatal: No se pudo crear el archivo dataset.csv.\n");
        return 1;
    }
    
    // Imprimir cabecera CSV
    fprintf(file, "ID,Nombre_Completo,Sexo,Edad,Peso_kg,Altura_cm\n");
    
    // Inicializar semilla de aleatoriedad
    srand(time(NULL));
    
    for (long long i = 1; i <= n; i++) {
        char sexo = (rand() % 2 == 0) ? 'M' : 'F';
        
        const char* nombre;
        if (sexo == 'M') {
            nombre = nombres_h[rand() % 10];
        } else {
            nombre = nombres_m[rand() % 10];
        }
        
        const char* apellido1 = apellidos[rand() % 12];
        const char* apellido2 = apellidos[rand() % 12];
        
        int edad = 18 + rand() % (90 - 18 + 1);
        double peso = rand_uniform(50.0, 110.0);
        int altura = 150 + rand() % (200 - 150 + 1);
        
        // Escribir el registro en el archivo de texto
        fprintf(file, "%lld,%s %s %s,%c,%d,%.1f,%d\n", i, nombre, apellido1, apellido2, sexo, edad, peso, altura);
        
        // Progreso para el usuario
        if (i % 1000000 == 0) {
            printf("... %lld registros generados.\n", i);
        }
    }
    
    fclose(file);
    printf("¡Generación completada a la velocidad de C! Datos guardados en 'dataset.csv'.\n");
    
    return 0;
}
